### PR TITLE
Simplified tally scoring

### PR DESCRIPTION
**Pull request by smharper**
This PR is an attempt to remove a lot of duplication from tally.F90.  I've introduced a new subroutine, `score_general` which handles the `select case(score_bin)` logic for every type of tally.

I haven't seen any large difference in runtime, but I also don't have any big tally tests.  Does anyone else have a problem with a ton of tallies that I can borrow for runtime testing?

There should be no difference as far as the user is concerned except for tracklength total_yn scores.  This PR should affect tracklength total_yn in the same way as #369 (which means that PR is superfluous if this one is accepted)

### CODE REVIEW

File: src/tally.F90 [hunk 1]
```diff
@@ -27,524 +27,573 @@ module tally
...
+    real(8) :: score_               ! analog tally score
     real(8) :: macro_total          ! material macro total xs
     real(8) :: macro_scatt          ! material macro scatt xs
-    logical :: found_bin            ! scoring bin found?
-    type(TallyObject), pointer, save :: t => null()
-!$omp threadprivate(t)
+    real(8) :: uvw(3)               ! particle direction
+    type(Material),    pointer, save :: mat => null()
+    type(Reaction),    pointer, save :: rxn => null()
+!$omp threadprivate(t, mat, rxn)
```
Comment (smharper) [inline src/tally.F90]: I'm not familiar with OpenMP so I'd appreciate some help with this.  It looks like `t` shouldn't be declared threadprivate, but I get a test_confidence_intervals failure if I remove it from the threadprivate declaration.

**Comment by smharper** [comment]
@wbinventor provided me with a very good stress test.  Looks like this PR is 50% slower than the develop so I'm going to close it.  Back to the drawing board.

**Comment by smharper** [comment]
Just kidding.  I'm reopening this PR.  I reran the tests and found no measurable performance drop.  It looks like my control last time found some quirk that made OpenMC run really fast.  I'd sure like to know how to repeat it.

**Comment by paulromano** [comment]
Fails to build with OpenMP

**Comment by smharper** [comment]
I think I've resolved my OpenMP issues.

File: src/tally.F90 [hunk 2]
```diff
@@ -27,524 +27,580 @@ module tally
...
-  subroutine score_analog_tally(p)
-
-    type(Particle), intent(in) :: p
-
-    integer :: i
-    integer :: i_tally
-    integer :: j                    ! loop index for scoring bins
-    integer :: k                    ! loop index for nuclide bins
+  subroutine score_general(p, t, start_index, filter_index, i_nuclide, &
+       &atom_density, flux)
```
Comment (paulromano) [inline src/tally.F90]: ampersand on second line is unnecessary (only needed when continuing a string)

File: src/tally.F90 [hunk 3]
```diff
@@ -27,524 +27,580 @@ module tally
...
+          ! Since only scattering events make it here, again we can use
+          ! the weight entering the collision as the estimator for the
+          ! reaction rate
+          score = p % last_wgt
+
+        else if (t % estimator == ESTIMATOR_TRACKLENGTH) then
+          ! Note SCORE_SCATTER_N not available for tracklength.
+          if (i_nuclide > 0) then
+            score = (micro_xs(i_nuclide) % total &
+                 &- micro_xs(i_nuclide) % absorption) * atom_density * flux
```
Comment (paulromano) [inline src/tally.F90]: Remove ampersand on second line

File: src/tally.F90 (hunk 2 above)
Comment (smharper) [inline src/tally.F90]: The ampersands are removed

**Comment by paulromano** [comment]
I'm seeing very similar performance as well. One could argue that there is some extra function invocation overhead, but in practice it shouldn't really matter. Good job getting rid of 700 lines of code!

**Pull request merged**